internal/providersdkv2/provider.go

```diff
@@ -111,24 +131,33 @@ func configure(p *schema.Provider) func(context.Context, *schema.ResourceData) (
 			diags = isHCPOperational()
 		}
 
-		// Sets up HCP SDK client.
-		userAgent := p.UserAgent("terraform-provider-hcp", version.ProviderVersion)
+		clientConfig := clients.ClientConfig{
+			ClientID:       d.Get("client_id").(string),
+			ClientSecret:   d.Get("client_secret").(string),
+			CredentialFile: d.Get("credential_file").(string),
+			SourceChannel:  p.UserAgent("terraform-provider-hcp", version.ProviderVersion),
+		}
 
-		clientID := d.Get("client_id").(string)
-		if clientID == "" {
-			clientID = os.Getenv("HCP_CLIENT_ID")
+		// Check the environment for an HCP Service Principal
+		if clientConfig.ClientID == "" {
+			clientConfig.ClientID = os.Getenv("HCP_CLIENT_ID")
 		}
-		clientSecret := d.Get("client_secret").(string)
-		if clientSecret == "" {
-			clientSecret = os.Getenv("HCP_CLIENT_SECRET")
+		if clientConfig.ClientSecret == "" {
+			clientConfig.ClientSecret = os.Getenv("HCP_CLIENT_SECRET")
 		}
 
-		client, err := clients.NewClient(clients.ClientConfig{
-			ClientID:       clientID,
-			ClientSecret:   clientSecret,
-			CredentialFile: d.Get("credential_file").(string),
-			SourceChannel:  userAgent,
-		})
+		// Read the workload_identity configuration
+		if v, ok := d.GetOk("workload_identity"); ok && len(v.([]interface{})) == 1 && v.([]interface{})[0] != nil {
+			wi := v.([]interface{})[0].(map[string]interface{})
+			if tf, ok := wi["token_file"].(string); ok && tf != "" {
+				clientConfig.WorloadIdentityTokenFile = tf
+			}
+			if rn, ok := wi["resource_name"].(string); ok && rn != "" {
+				clientConfig.WorkloadIdentityResourceName = rn
+			}
```

Comment:
I haven't tested it, but I suspect that this could also be done with `mapstructure`, which already is a dependency of the provider. It would be a bit easier to read.

In `client.go`:
```go
WorloadIdentityTokenFile     string `mapstructure:"token_file"`
WorkloadIdentityResourceName string `mapstructure:"resource_name"`
```

Here:
```go
err := mapstructure.Decode(v.([]interface{})[0], clientConfig)
if err != nil {
  diags = append(diags, diag.Errorf("unable to parse workload identity config: %v", err)...)
  return nil, diags
}
```